File: app/src/main/java/com/example/myapplication/feature/device/PacketProcessingUpdateBatcher.kt
package com.example.myapplication.feature.device

class PacketProcessingUpdateBatcher(
    private val dispatchIntervalMillis: Long,
    private val schedule: (Runnable, Long) -> Unit,
    private val cancel: (Runnable) -> Unit,
    private val dispatch: (Long, PacketProcessingUpdate) -> Unit,
) {
    private val lock = Any()
    private val dispatchRunnable = Runnable { flushScheduled() }

    private var generation = 0L
    private var scheduledGeneration: Long? = null
    private var pendingUpdate: PendingPacketProcessingUpdate? = null

    fun submit(update: PacketProcessingUpdate) {
        if (dispatchIntervalMillis <= 0L) {
            val currentGeneration = synchronized(lock) { generation }
            dispatch(currentGeneration, update)
            return
        }

        synchronized(lock) {
            val pending = pendingUpdate ?: PendingPacketProcessingUpdate()
            pending.merge(update)
            pendingUpdate = pending
            if (scheduledGeneration == null) {
                scheduledGeneration = generation
                schedule(dispatchRunnable, dispatchIntervalMillis)
            }
        }
    }

    fun flushNow() {
        if (dispatchIntervalMillis <= 0L) return

        synchronized(lock) {
            cancel(dispatchRunnable)
            scheduledGeneration = null
            val updateToDispatch = pendingUpdate?.toImmutable()
            pendingUpdate = null
            if (updateToDispatch != null) {
                dispatch(generation, updateToDispatch)
            }
        }
    }

    fun clearPending() {
        synchronized(lock) {
            generation++
            scheduledGeneration = null
            pendingUpdate = null
            cancel(dispatchRunnable)
        }
    }

    private fun flushScheduled() {
        synchronized(lock) {
            val targetGeneration = scheduledGeneration
            scheduledGeneration = null
            if (targetGeneration == null || targetGeneration != generation) {
                return
            }

            val updateToDispatch = pendingUpdate?.toImmutable()
            pendingUpdate = null
            if (updateToDispatch != null) {
                dispatch(generation, updateToDispatch)
            }
        }
    }

    fun isGenerationCurrent(expectedGeneration: Long): Boolean {
        return synchronized(lock) { generation == expectedGeneration }
    }

    private class PendingPacketProcessingUpdate {
        private val chartSamplesByStream = linkedMapOf<ChartStreamKey, MutableList<ChartPoint>>()
        private val diagnosticEvents = mutableListOf<PacketDiagnosticEvent>()

        private var packetsReceived = 0L
        private var packetsLost = 0L
        private var packetsRejected = 0L
        private var timerRegressionRejects = 0L
        private var fragmentsReceived = 0L
        private var rawBytesReceived = 0L
        private var lastPacketIssue: String? = null
        private var rejectionBreakdown: String? = null

        fun merge(update: PacketProcessingUpdate) {
            packetsReceived = update.packetsReceived
            packetsLost = update.packetsLost
            packetsRejected = update.packetsRejected
            timerRegressionRejects = update.timerRegressionRejects
            fragmentsReceived = update.fragmentsReceived
            rawBytesReceived = update.rawBytesReceived
            if (update.lastPacketIssue != null) {
                lastPacketIssue = update.lastPacketIssue
            }
            if (update.rejectionBreakdown != null) {
                rejectionBreakdown = update.rejectionBreakdown
            }

            update.chartSamplesByStream.forEach { (streamKey, points) ->
                if (points.isEmpty()) return@forEach
                chartSamplesByStream.getOrPut(streamKey) { mutableListOf() }.addAll(points)
            }
            diagnosticEvents += update.diagnosticEvents
        }

        fun toImmutable(): PacketProcessingUpdate {
            return PacketProcessingUpdate(
                packetsReceived = packetsReceived,
                packetsLost = packetsLost,
                packetsRejected = packetsRejected,
                timerRegressionRejects = timerRegressionRejects,
                fragmentsReceived = fragmentsReceived,
                rawBytesReceived = rawBytesReceived,
                chartSamplesByStream = chartSamplesByStream.mapValues { (_, points) -> points.toList() },
                lastPacketIssue = lastPacketIssue,
                rejectionBreakdown = rejectionBreakdown,
                diagnosticEvents = diagnosticEvents.toList(),
            )
        }
    }
}

class BatchingPacketCaptureController(
    private val delegate: PacketCaptureController,
    private val batcher: PacketProcessingUpdateBatcher,
) : PacketCaptureController {
    override fun submit(packetFragment: ByteArray): PacketSubmitResult = delegate.submit(packetFragment)

    override fun recordDiagnosticEvent(type: PacketDiagnosticType, message: String) {
        delegate.recordDiagnosticEvent(type, message)
    }

    override fun stopCapture() {
        delegate.stopCapture()
        batcher.clearPending()
    }

    override fun updateSelection(sensorType: Int, channel: Int) {
        delegate.updateSelection(sensorType, channel)
    }

    override fun resetSession() {
        delegate.resetSession()
        batcher.clearPending()
    }

    override fun flush() {
        batcher.flushNow()
        delegate.flush()
    }

    override fun currentFile() = delegate.currentFile()

    override fun currentPacketFile() = delegate.currentPacketFile()

    override fun currentRawFile() = delegate.currentRawFile()

    override fun currentLogFile() = delegate.currentLogFile()

    override fun close() {
        delegate.stopCapture()
        batcher.clearPending()
        delegate.close()
    }
}

File: ios/Woona/Capture/PacketProcessingUpdateBatcher.swift
import Foundation

@MainActor
final class PacketProcessingUpdateBatcher {
    private let dispatchIntervalNanos: UInt64
    private let dispatch: @MainActor (PacketProcessingUpdate) -> Void

    private var pendingUpdate: PendingPacketProcessingUpdate?
    private var scheduledTask: Task<Void, Never>?
    private var generation: UInt64 = 0

    init(
        dispatchIntervalMillis: UInt64 = 100,
        dispatch: @escaping @MainActor (PacketProcessingUpdate) -> Void
    ) {
        self.dispatchIntervalNanos = dispatchIntervalMillis * 1_000_000
        self.dispatch = dispatch
    }

    func submit(_ update: PacketProcessingUpdate) {
        if dispatchIntervalNanos == 0 {
            dispatch(update)
            return
        }

        var pending = pendingUpdate ?? PendingPacketProcessingUpdate()
        pending.merge(update)
        pendingUpdate = pending

        guard scheduledTask == nil else { return }

        let targetGeneration = generation
        let delay = dispatchIntervalNanos
        scheduledTask = Task { [weak self] in
            do {
                try await Task.sleep(nanoseconds: delay)
            } catch {
                return
            }
            self?.flushScheduled(generation: targetGeneration)
        }
    }

    func flushNow() {
        generation += 1
        scheduledTask?.cancel()
        scheduledTask = nil

        guard let update = pendingUpdate?.toUpdate() else { return }
        pendingUpdate = nil
        dispatch(update)
    }

    func clearPending() {
        generation += 1
        scheduledTask?.cancel()
        scheduledTask = nil
        pendingUpdate = nil
    }

    private func flushScheduled(generation targetGeneration: UInt64) {
        guard targetGeneration == generation else { return }
        scheduledTask = nil
        guard let update = pendingUpdate?.toUpdate() else { return }
        pendingUpdate = nil
        dispatch(update)
    }
}

private struct PendingPacketProcessingUpdate {
    private var packetsReceived: UInt64 = 0
    private var packetsLost: UInt64 = 0
    private var packetsRejected: UInt64 = 0
    private var timerRegressionRejects: UInt64 = 0
    private var fragmentsReceived: UInt64 = 0
    private var rawBytesReceived: UInt64 = 0
    private var chartSamplesByStream: [ChartStreamKey: [ChartPoint]] = [:]
    private var lastPacketIssue: String?
    private var rejectionBreakdown: String?
    private var diagnosticEvents: [PacketDiagnosticEvent] = []

    mutating func merge(_ update: PacketProcessingUpdate) {
        packetsReceived = update.packetsReceived
        packetsLost = update.packetsLost
        packetsRejected = update.packetsRejected
        timerRegressionRejects = update.timerRegressionRejects
        fragmentsReceived = update.fragmentsReceived
        rawBytesReceived = update.rawBytesReceived

        for (key, points) in update.chartSamplesByStream where !points.isEmpty {
            chartSamplesByStream[key, default: []].append(contentsOf: points)
        }

        if let issue = update.lastPacketIssue {
            lastPacketIssue = issue
        }
        if let breakdown = update.rejectionBreakdown {
            rejectionBreakdown = breakdown
        }
        diagnosticEvents.append(contentsOf: update.diagnosticEvents)
    }

    func toUpdate() -> PacketProcessingUpdate {
        PacketProcessingUpdate(
            packetsReceived: packetsReceived,
            packetsLost: packetsLost,
            packetsRejected: packetsRejected,
            timerRegressionRejects: timerRegressionRejects,
            fragmentsReceived: fragmentsReceived,
            rawBytesReceived: rawBytesReceived,
            chartSamplesByStream: chartSamplesByStream,
            lastPacketIssue: lastPacketIssue,
            rejectionBreakdown: rejectionBreakdown,
            diagnosticEvents: diagnosticEvents
        )
    }
}
